src/nvim/tui/input.c

```diff
@@ -692,10 +691,19 @@ static void handle_raw_buffer(TermInput *input, bool force)
     }
     // Push through libtermkey (translates to "<keycode>" strings, etc.).
     RBUFFER_UNTIL_EMPTY(input->read_stream.buffer, ptr, len) {
-      size_t consumed = termkey_push_bytes(input->tk, ptr, MIN(count, len));
-      // termkey_push_bytes can return (size_t)-1, so it is possible that
-      // `consumed > rbuffer_size(input->read_stream.buffer)`, but since tk_getkeys is
-      // called soon, it shouldn't happen.
+      size_t size = MIN(count, len);
+      if (size > termkey_get_buffer_remaining(input->tk)) {
+        // We are processing a very long escape sequence. Increase termkey's
+        // internal buffer size. We don't handle out of memory situations so
+        // assert the result
+        const size_t bufsize = termkey_get_buffer_size(input->tk);
+        assert(termkey_set_buffer_size(input->tk, 2 * bufsize));
```

Comment:
Doubling the buffer size is just a heuristic. If the input is very large we will have to reallocate multiple times. But this should not happen very often (the _only_ case we ever see this is for large OSC 52 pastes, which are infrequent).